src/serial.rs

```diff
@@ -61,49 +61,55 @@ pub enum Error {
     _Extensible,
 }
 
-pub trait Pins<USART> {}
+pub trait TxPin<USART> {}
+pub trait RxPin<USART> {}
 
-#[cfg(any(feature = "stm32f030", feature = "stm32f042"))]
-impl Pins<stm32::USART1> for (gpioa::PA9<Alternate<AF1>>, gpioa::PA10<Alternate<AF1>>) {}
-#[cfg(any(feature = "stm32f030", feature = "stm32f042"))]
-impl Pins<stm32::USART1> for (gpiob::PB6<Alternate<AF0>>, gpiob::PB7<Alternate<AF0>>) {}
-#[cfg(any(feature = "stm32f030", feature = "stm32f042"))]
-impl Pins<stm32::USART1> for (gpioa::PA9<Alternate<AF1>>, gpiob::PB7<Alternate<AF0>>) {}
-#[cfg(any(feature = "stm32f030", feature = "stm32f042"))]
-impl Pins<stm32::USART1> for (gpiob::PB6<Alternate<AF0>>, gpioa::PA10<Alternate<AF1>>) {}
+macro_rules! usart_pins {
+    ($($USART:ident: ($tx:ty, $rx:ty),)+) => {
+        $(
+            impl TxPin<stm32::$USART> for $tx {}
+            impl RxPin<stm32::$USART> for $rx {}
+        )+
+    }
+}
 
+#[cfg(any(feature = "stm32f030", feature = "stm32f042"))]
+usart_pins! {
+    USART1: (gpioa::PA9<Alternate<AF1>>, gpioa::PA10<Alternate<AF1>>),
+    USART1: (gpiob::PB6<Alternate<AF0>>, gpiob::PB6<Alternate<AF0>>),
+}
 #[cfg(feature = "stm32f030x6")]
-impl Pins<stm32::USART1> for (gpioa::PA2<Alternate<AF1>>, gpioa::PA3<Alternate<AF1>>) {}
-
-#[cfg(any(
-    feature = "stm32f042",
-    feature = "stm32f030x8",
-    feature = "stm32f030xc",
-))]
-impl Pins<stm32::USART2> for (gpioa::PA2<Alternate<AF1>>, gpioa::PA3<Alternate<AF1>>) {}
-#[cfg(any(
-    feature = "stm32f042",
-    feature = "stm32f030x8",
-    feature = "stm32f030xc",
-))]
-impl Pins<stm32::USART2> for (gpioa::PA2<Alternate<AF1>>, gpioa::PA15<Alternate<AF1>>) {}
-#[cfg(any(
-    feature = "stm32f042",
-    feature = "stm32f030x8",
-    feature = "stm32f030xc",
-))]
-impl Pins<stm32::USART2> for (gpioa::PA14<Alternate<AF1>>, gpioa::PA15<Alternate<AF1>>) {}
+usart_pins! {
+    USART1: (gpioa::PA2<Alternate<AF1>>, gpioa::PA3<Alternate<AF1>>),
+    USART1: (gpioa::PA14<Alternate<AF1>>, gpioa::PA15<Alternate<AF1>>),
+}
 #[cfg(any(
     feature = "stm32f042",
     feature = "stm32f030x8",
     feature = "stm32f030xc",
 ))]
-impl Pins<stm32::USART2> for (gpioa::PA14<Alternate<AF1>>, gpioa::PA3<Alternate<AF1>>) {}
+usart_pins! {
+    USART2: (gpioa::PA2<Alternate<AF1>>, gpioa::PA3<Alternate<AF1>>),
+    USART2: (gpioa::PA14<Alternate<AF1>>, gpioa::PA15<Alternate<AF1>>),
+}
+#[cfg(feature = "stm32f030xc")]
+usart_pins! {
+    // TODO WTF look at this again, in the datasheet PB10 is both tx and rx
+    // USART3: (gpiob::PB10, AF4, gpiob::PA11, AF4),
+    USART3: (gpioc::PC4<Alternate<AF1>>, gpioc::PC5<Alternate<AF1>>),
```

Comment:
Instead of specifying pairs wouldn't it make more sense to actually single out the TX/RX functions and let the user chose any possible combination? Using this macro in this way means you're passing on a few possible combinations, e.g. you use also use PC10 and PC5 for USART3, or PC4 and PC11.

Reply:
Thats the whole purpose of the separate Rx and Tx traits, the combinations are done by the templates. The only reason i defined them pairwise is because that what ST seems to have done and it seemed cleaner this way.

Reply:
@HarkonenBade That sounds like a good idea.

Reply:
@HarkonenBade Agree. We could also reuse that for the spi, where the pin combinations aren't that clear.

Reply:
Done, if @david-sawatzke could pull the fresh commit from https://github.com/HarkonenBade/stm32f0xx-hal/tree/stm32f030_v2 that would be great.

Reply:
I took some slight liberties with making the macro syntax a a bit more verbose, but I think it makes it a lot clearer what is going on.

Reply:
@HarkonenBade Yeah, that's better. The syntax also seems more "rusty".